rfcs/000-third-party-software-installation.md

```diff
@@ -0,0 +1,109 @@
+- Topic Name: `third-party-software-installation`
+- Start Date: 2022-03-22
+- RFC PR: [vlang/rfcs#0000](https://github.com/vlang/rfcs/pull/22)
+  <!-- - V Issue: [vlang/v#00000](https://github.com/vlang/v/issues/00000) -->
+
+# Summary
+
+Allow installation of third-party V software.
+This goes beyond just placing the source code in `VMODULES`.
+It allows building the software from source locally and adding it to path for the user to uses from anywhere.
+
+Additionally, we propose the use of `~/.v` and the relevant subfolders (like `~/.v/modules`, `~/.v/bin`) to store the source code, built binaries, etc.
+
+# Motivation
+
+Currently, V only allows downloading the source code for V modules in the `VMODULES` directory through the `v install` command.
+This is good for making modules globally available. There is currently no streamlined way of _installing_ or _fetching_ or _obtaining_ third-party software written in V.
+
+# Guide-level explanation
+
+Take [_klonol_](https://github.com/hungrybluedev/klonol) for example.
+It is a command-line tool written in V that helps to maintain a local copy of all git repositories on GitHub and Gitea.
+If you want to install it, you have to clone the source yourself, run `v build.vsh` and then add the `bin` folder to `PATH`.
+A nice thing to have is a `v fetch <repo>` or `v obtain <repo>` command (or something else, please comment) that will clone the repo and build it for you.
+
+Only once, like when V is install, for example, the user needs to add `~/.v/bin` to `PATH`.
+This will make all third-party applications installed using `v fetch <repo>` available from anywhere.
+
+# Reference-level explanation
+
+It will piggyback off of `v install`.
+The act of installing the software is done through the following additional steps:
+
+1. Running `v build.vsh` or some other specified build script in the root of the repo.
+2. This will create an executable, along with all other
+   relevant files (like templates, etc) in the `bin` directory of the repo. Ideally, the resources will be embedded in the executable to allow for only one executable to be enough.
+3. The executable will be added to `~/.v/bin` for the user to call from anywhere.
+   TODO.
+   An additional **suggestion** is to have the `~/.vmodules` directory be replaced in favour of `~/.v/modules`.
+
+If there is another script (other than `build.vsh`) that generates the executable
+
+A sample `build.vsh` script is:
+
+```text
+println('Removing old artifacts...')
```

Comment:
Something like a setup function like Python has it would be good. Then we have a more standardized way of building a package.

Reply:
@JalonSolov had already suggested something like a post-installation script or set of commands.

Reply:
If it's a script, it will always have too many possibilities which result in security risks.

Reply:
I'm afraid we'll not go away with non-scripting capabilities. Many tried, all of them failed. Even Python's setup function is not enough.

Actually we really need the whole machinery as e.g. [PKGBUILD](https://wiki.archlinux.org/title/PKGBUILD )s do (anyone telling me we don't was not distribution's package maintainer for long enough or didn't maintain any complex package).

Please just don't reinvent the wheel and go for what e.g. PKGBUILDs do (with some [minor twists & goodies](https://musescore.org/en/developers-handbook/packaging-different-operating-systems/packaging-linux-bsd-systems-maintainer-wis ) but especially **tight integration with the *existing* V build system aka `vpm`**).

Whatever you'll do in the end, note that there should be seamless support for both binary as well as source files distribution (i.e. a way to separately describe what to build during packaging and what to build at the end-user place but still before the actual installation).

I'm not recommending rpm spec nor debian packages nor npm-like stuff nor pip-like nor CPAN-like etc. because they're often not strictly single-file oriented, have actually historically a lot of unnecessary burden (mainly stemming from the fact that they so much wanted it to look declarative until they ended up with ad-hoc syntax extensions reaching to hells like multiple macro preprocessing steps :sob:, bloat, etc.) and are still not universal nor lightweight. Lesson learned: make the package description into a pure **existing** programming language (in our case *V*).

But I'd still advise to get some inspiration from other notable packaging systems: [Alpine apk-tools]( https://wiki.alpinelinux.org/wiki/Apk_spec ), [Nix](https://nixos.wiki/wiki/Packaging/Binaries ), [Void's xbps](https://github.com/void-linux/xbps ), [homebrew's brew](https://brew.sh/ ), [BSD's "Ports" pkg](https://docs.freebsd.org/en/books/handbook/ports/ ), [exherbo's cave from Paludis](https://paludis.exherbo.org/clients/cave.html ), [funtoo's emerge from Portage](https://www.funtoo.org/Install/Introducing_Portage ) in this order (they all shall support both source & binary packages distribution in one way or another).